File: 2D-Arrays/20/islandsBfs.py
class Solution:
    def __init__(self):
        self.directions = [
            [-1, 0],
            [0, 1],
            [1, 0],
            [0, -1]
        ]

    def numIslands(self, grid: List[List[str]]) -> int:
        count = 0
        seen = [[False]*len(grid[0]) for _ in range(len(grid))]
        # Sequential search
        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if grid[row][col] == '1':
                    count += 1
                    self.convertOldIslands(row, col, grid, seen)
        return count

    # Using dfs to convert all old islands to 0
    def convertOldIslands(self, row, col, grid, seen):
        from collections import deque
        q = deque()
        q.append([row, col])
        while len(q) > 0:
            row, col = q.popleft()
            if row < 0 or row >= len(grid) or col < 0 or col >= len(grid[0]) or seen[row][col] or grid[row][col] == '0':
                continue
            grid[row][col] = '0'
            seen[row][col] = True
            for direction in self.directions:
                q.append([row+direction[0], col+direction[1]])

File: 2D-Arrays/20/author/islandsBfs.js
const directions = [
  [-1, 0], // up
  [0, 1], // right
  [1, 0], // down
  [0, -1], // left
];

var numIslands = function (grid) {
  if (grid.length === 0) return 0;
  let count = 0;

  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[0].length; j++) {
      if (grid[i][j] === '1') {
        count++;
        convertOldIslands(i, j, grid);
      }
    }
  }
  return count;
};

var convertOldIslands = function (i, j, grid) {
  const q = [[i, j]];
  grid[i][j] = '0';
  while (q.length > 0) {
    [i, j] = q.shift();
    for (let d = 0; d < directions.length; d++) {
      const currentDirection = directions[d];
      const nextRow = i + currentDirection[0];
      const nextCol = j + currentDirection[1];
      if (
        nextRow < 0 ||
        nextRow >= grid.length ||
        nextCol < 0 ||
        nextCol >= grid[0].length
      )
        continue;
      if (grid[nextRow][nextCol] === '1') {
        q.push([nextRow, nextCol]);
        grid[nextRow][nextCol] = '0';
      }
    }
  }
};
